packages/did/src/DidBatcher/FullDidUpdateBuilder.ts

```diff
@@ -0,0 +1,555 @@
+ */
+
+import type { Extrinsic } from '@polkadot/types/interfaces'
+import { ApiPromise } from '@polkadot/api'
+
+import type {
+  DidEncryptionKey,
+  KeystoreSigner,
+  IIdentity,
+  IDidIdentifier,
+  DidVerificationKey,
+  NewDidVerificationKey,
+  NewDidEncryptionKey,
+  DidServiceEndpoint,
+  DidKey,
+  SubmittableExtrinsic,
+} from '@kiltprotocol/types'
+import { KeyRelationship } from '@kiltprotocol/types'
+import { SDKErrors } from '@kiltprotocol/utils'
+
+import { FullDidDetails } from '../DidDetails/FullDidDetails.js'
+import { increaseNonce } from '../DidDetails/FullDidDetails.utils.js'
+import {
+  formatPublicKey,
+  generateDidAuthenticatedTx,
+  queryNonce,
+} from '../Did.chain.js'
+import { getSignatureAlgForKeyType } from '../Did.utils.js'
+
+import { FullDidBuilder } from './FullDidBuilder.js'
+import { deriveChainKeyId } from './FullDidBuilder.utils.js'
+
+export type FullDidUpdateBuilderCreationDetails = {
+  authenticationKey: DidVerificationKey
+  identifier: IDidIdentifier
+  keyAgreementKeys?: DidEncryptionKey[]
+  assertionKey?: DidVerificationKey
+  delegationKey?: DidVerificationKey
+  serviceEndpoints?: DidServiceEndpoint[]
+}
+
+export type FullDidUpdateHandler = (
+  didUpdateExtrinsicBatch: SubmittableExtrinsic
+) => Promise<void>
+
+// TODO: replace manual extrinsic creation with DID chain functions when those will take an api object as parameter, so that they do not have to be async.
+
+/**
+ * A builder to batch multiple changes before a DID update.
+ */
+export class FullDidUpdateBuilder extends FullDidBuilder {
+  protected identifier: IDidIdentifier
+  protected batch: Extrinsic[] = []
+
+  protected oldAuthenticationKey: DidVerificationKey
+  protected newAuthenticationKey: NewDidVerificationKey | undefined = undefined
+
+  // Old key agreement key as {id -> details}
+  protected oldKeyAgreementKeys: Map<
+    DidEncryptionKey['id'],
+    Omit<DidEncryptionKey, 'id'>
+  > = new Map()
+
+  // Key agreement keys to delete, by their ID.
+  protected keyAgreementKeysToDelete: Set<DidEncryptionKey['id']> = new Set()
+
+  // Old assertion key, if present
+  protected oldAssertionKey: DidVerificationKey | undefined = undefined
+  // Old delegation key, if present
+  protected oldDelegationKey: DidVerificationKey | undefined = undefined
+
+  // Old service endpoints as {id -> details}
+  protected oldServiceEndpoints: Map<
+    DidServiceEndpoint['id'],
+    Omit<DidServiceEndpoint, 'id'>
+  > = new Map()
+
+  // Service endpoints to delete, by their ID.
+  protected serviceEndpointsToDelete: Set<DidServiceEndpoint['id']> = new Set()
+
+  /**
+   * Initialize a DID update with the information contained in the provided full DID.
+   *
+   * @param api The [[ApiPromise]] object to encode/decoded types as needed.
+   * @param details The [[FullDidDetails]] object.
+   */
+  public constructor(api: ApiPromise, details: FullDidDetails) {
+    super(api)
+    ;({
```

Comment:
C: never saw this way of assigning variables to fields. I would say it is harder to read, than just assigning them the classic way.

Reply:
https://github.com/KILTprotocol/sdk-js/pull/473/commits/1fa1781e199bb913e78050ce248250cd31629f30